Доработана функциональность по техническому заданию 10 спринта

asdbl opened the pull request.

asdbl committed:
Добавить User и Film.
Добавить UserController и FilmController

asdbl committed:
Изменить checkstyle тесты.

asdbl committed:
Изменить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

asdbl committed:
Исправить checkstyle тесты.

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,72 @@
+@RestController
+@RequestMapping("/users")
+public class UserController {
+    private final HashMap<Long, User> users = new HashMap<>();
```
Раз ты познакомился с интерфейсами и полиморфизмом, может быть стоит теперь использовать интерфейс `Map`, чтобы не привязывать к конкретной его реализации, а работать лишь с предоставляемым API) Все места не отмечал, но думаю, что ты разберешься)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,77 @@
 public class FilmController {
+    public static final Integer MAX_DESCRIPTION_LENGTH = 200;
+    public static final LocalDate MIN_RELEASE_DATE = LocalDate.of(1895, 12, 28);
+    private final HashMap<Long, Film> films = new HashMap<>();
```
Раз ты познакомился с интерфейсами и полиморфизмом, может быть стоит теперь использовать интерфейс `Map`, чтобы не привязывать к конкретной его реализации, а работать лишь с предоставляемым API) Все места не отмечал, но думаю, что ты разберешься)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,77 @@
+    public static final Integer MAX_DESCRIPTION_LENGTH = 200;
+    public static final LocalDate MIN_RELEASE_DATE = LocalDate.of(1895, 12, 28);
+    private final HashMap<Long, Film> films = new HashMap<>();
+    private static final Logger LOG = LoggerFactory.getLogger(FilmController.class);
```
Эту логику можно было бы реализовать с использованием аннотации `@Slf4j` над классом-контроллером)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,72 @@
+@RequestMapping("/users")
+public class UserController {
+    private final HashMap<Long, User> users = new HashMap<>();
+    private static final Logger LOG = LoggerFactory.getLogger(UserController.class);
```
Эту логику можно было бы реализовать с использованием аннотации `@Slf4j` над классом-контроллером)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,72 @@
+public class UserController {
+    private final HashMap<Long, User> users = new HashMap<>();
+    private static final Logger LOG = LoggerFactory.getLogger(UserController.class);
+    private static Long id = 0L;
```
Не требуется делать статическим, так как класс `UserController` является бином и будет создан в единственном экземпляре)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,77 @@
+    public static final LocalDate MIN_RELEASE_DATE = LocalDate.of(1895, 12, 28);
+    private final HashMap<Long, Film> films = new HashMap<>();
+    private static final Logger LOG = LoggerFactory.getLogger(FilmController.class);
+    private static Long id = 0L;
```
Не требуется делать статическим, так как класс `FilmController` является бином и будет создан в единственном экземпляре)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,23 @@
 @Setter
 public class Film {
+    private Long id;
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как проверка на null входит в `@NotBlank`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,23 @@
  */
 @Getter
 @Setter
 public class Film {
```
Над описанием можно было бы использовать `@Size`, а для даты релиза использовать @NotNull, чтобы запретить не передавать ее) Для продолжительности `@Positive` или `@Min(1)`, чтобы запретить передавать отрицательное значение, но также потребуется использовать `@NotNull` либо же потребуется изменить тип на примитивный, так как обе допускают `null`) Чтобы эти аннотации стали учитываться, следует добавить перед параметром `Film film`, в методах создания и обновления, аннотацию `@Valid`. Аналогично и с пользователем)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,23 @@
+    @NotBlank
+    private String name;
+    private String description;
+    private LocalDate releaseDate;
```
(предложение) Если будет интересно, то для валидации даты фильма можно реализовать собственную аннотацию валидации, чтобы придерживаться более декларативного подхода)
https://devcolibri.com/spring-mvc-кастомная-аннотация-для-валидации/

asdbl replied:
Долго не мог понять почему у меня не работает кастомная аннотация, оказалось, что нужно использовать jakarta вместо javax)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,20 @@
+    private Long id;
+    @Email
+    private String email;
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как проверка на null входит в `@NotBlank`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,20 @@
+    @NotBlank
+    private String login;
+    private String name;
+    private LocalDate birthday;
```
Для валидации даты рождения есть очень удобная аннотация -  `@PastOrPresent`, но она должна использовать в связке с `@NotNull`, так как она допускает `null` за корректное значение)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,20 @@
+public class User {
+    private Long id;
+    @Email
+    private String email;
```
Для валидации почты следует использовать `@Email` в связке с `@NotEmpty`, так как `@Email` допускает `null` и пустую строку за корректные значения)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,20 @@
+    private String email;
+    @NotNull
+    @NotBlank
+    private String login;
```
(предложение) Так как строка с логином не может быть `null` и состоять только из пробелов, хорошо было бы использовать аннотацию `@NotBlank`, что уже есть. Также строка не может содержать пробелов - для этой цели было бы удобно использовать аннотацию `@Pattern`
https://www.geeksforgeeks.org/spring-mvc-regular-expression-validation/

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,72 @@
+        return user;
+    }
+
+    private static void userValidation(User user) {
```
Часть проверок можно было бы исключить, так как они реализованы с помощью аннотаций валидации) Кстати, тестирование валидации с помощью аннотаций должно осуществляться подобным образом
https://stackoverflow.com/questions/29069956/how-to-test-validation-annotations-of-a-class-using-junit

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,77 @@
+        return film;
+    }
+
+    private static void filmValidation(Film film) {
```
Часть проверок можно было бы исключить, так как они реализованы с помощью аннотаций валидации) Кстати, тестирование валидации с помощью аннотаций должно осуществляться подобным образом
https://stackoverflow.com/questions/29069956/how-to-test-validation-annotations-of-a-class-using-junit

TyutterinYakov replied:
Хорошо было бы придерживаться более принятого порядка во всех классах и сначала располагать поля - конструкторы - публичные методы(предоставляемый API), а затем уже приватные - внутренняя реализация). Таким образом будет проще извлекать информацию из кода и ориентироваться по нему

TyutterinYakov reviewed (commented):
Добрый день! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего дня! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌅

asdbl committed:
Заменить логику для логирования наа аннотацию.
Удалить аннотацию @NotNull, где используется @NotNull.
Добавить аннотацию @Size для описания фильма.
Добавить аннотацию @Positive для длительности фильма.
Добавить аннотацию @PastOrPresent для даты рождения пользователя.
Добавить аннотацию @Pattern для логина.
Удалить чсть проверок, которые реализовны с помощью аннотаций.

asdbl committed:
Исправить checkstyle.xml

asdbl committed:
Исправить checkstyle.xml

asdbl committed:
Исправить checkstyle.xml

asdbl committed:
Исправить checkstyle.xml

asdbl committed:
Please enter the commit message for your changes. Lines starting

asdbl committed:
Исправить checkstyle.

TyutterinYakov commented on pom.xml:
```diff
@@ -32,8 +32,33 @@
+		<dependency>
+			<groupId>jakarta.validation</groupId>
+			<artifactId>jakarta.validation-api</artifactId>
+			<version>3.1.0</version>
```
Действительно ли эта зависимость требуется?) Выше уже присутствует `spring-boot-starter-validation`, которая включает в себя все необходимое

TyutterinYakov commented on pom.xml:
```diff
@@ -32,8 +32,33 @@
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
+			<artifactId>hibernate-validator</artifactId>
+			<version>8.0.1.Final</version>
```
Действительно ли эта зависимость требуется?) Выше уже присутствует `spring-boot-starter-validation`, которая включает в себя все необходимое

TyutterinYakov commented on pom.xml:
```diff
@@ -32,8 +32,33 @@
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-validation</artifactId>
+			<version>3.2.4</version>
```
Указание версии у зависимостей `org.springframework.boot` - не требуется, так как она берется от родительского проекта, чтобы обеспечить полную совместимость)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,27 @@
+    private String name;
+    @Size(max = 200, message = "Мксимальня длина описания - 200 символов")
+    private String description;
+    @NotNull(message = "Дата выхода не может быть пустой")
```
Проверка на `null` уже реализована в кастомном валидаторе) Здесь `@NotNull` не требуется

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,27 @@
+    private Long id;
+    @NotBlank(message = "Имя не может быть пустым")
+    private String name;
+    @Size(max = 200, message = "Мксимальня длина описания - 200 символов")
```
Также хорошо было бы добавить `@NotBlank`, чтобы не допустить `null`, пустую или строку лишь из пробелов)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,27 @@
+    @NotNull(message = "Дата выхода не может быть пустой")
+    @ReleaseDate(year = 1895, month = 12, day = 28, message = "Дата выхода фильма не может быть раньше 1895-12-28")
+    private LocalDate releaseDate;
+    @NotNull(message = "Длительность фильма не может быть пустой")
```
`@NotNull` не требуется использовать над примитивным типом, так как они не могут иметь значение `null`

TyutterinYakov reviewed (approved):
Добрый день! Отметил оставшиеся небольшие недочеты, но думаю, что ты справишься с ними вне ревью) Успехов в обучении и прекрасного дня!🌅🎇

asdbl committed:
Удалить неиспользуемые зависимости.
Исправить аннотации в Film.

asdbl committed:
Исправить checkstyle.
